plan/new_physical_plan_builder.go

```diff
@@ -110,21 +110,60 @@ func (p *LogicalJoin) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile,
 	}
 	switch p.JoinType {
 	case SemiJoin, LeftOuterSemiJoin:
-		task, err = p.convert2SemiJoin()
+		task, err = p.convert2SemiJoin(prop)
 	default:
-		// TODO: We will consider smj and index look up join in the future.
-		task, err = p.convert2HashJoin()
+		if p.preferUseMergeJoin() {
+			task, err = p.convert2MergeJoin(prop)
+		} else {
+			// TODO: We will consider index look up join in the future.
+			task, err = p.convert2HashJoin(prop)
+		}
 	}
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	// Because hash join is executed by multiple goroutines, it will not propagate physical property any more.
-	// TODO: We will consider the problem of property again for parallel execution.
-	task = prop.enforceProperty(task, p.ctx, p.allocator)
 	return task, p.storeTaskProfile(prop, task)
 }
 
-func (p *LogicalJoin) convert2SemiJoin() (taskProfile, error) {
+func (p *LogicalJoin) preferUseMergeJoin() bool {
+	return p.preferMergeJoin && len(p.EqualConditions) == 1
+}
+
+func (p *LogicalJoin) convert2MergeJoin(prop *requiredProp) (taskProfile, error) {
+	lChild := p.children[0].(LogicalPlan)
+	rChild := p.children[1].(LogicalPlan)
+	mergeJoin := PhysicalMergeJoin{
+		JoinType:        p.JoinType,
+		EqualConditions: p.EqualConditions,
+		LeftConditions:  p.LeftConditions,
+		RightConditions: p.RightConditions,
+		OtherConditions: p.OtherConditions,
+	}.init(p.allocator, p.ctx)
+	mergeJoin.SetSchema(p.schema)
+	lJoinKey := p.EqualConditions[0].GetArgs()[0].(*expression.Column)
```

Comment:
I think you either need to take just one condition as join condition or push all remaining equal condition down to child plan. Merge join will assume all eq condition keys are sorted accordingly. But here if only one key is pushed as required property, 2nd, 3rd and other remaining keys might be unsorted for the same 1st key group.